src/mlpack/methods/fastmks/fastmks_rules.hpp

```diff
@@ -101,10 +124,28 @@ class FastMKSRules
   //! The query dataset.
   const typename TreeType::Mat& querySet;
 
-  //! The indices of the maximum kernel results.
-  arma::Mat<size_t>& indices;
-  //! The maximum kernels.
-  arma::mat& products;
+  //! Candidate represents a possible candidate point (value, index).
+  typedef std::pair<double, size_t> Candidate;
+
+  //! Compare two candidates based on the value.
+  struct CandidateCmp {
+    bool operator()(const Candidate& c1, const Candidate& c2)
+    {
+      return c1.first > c2.first;
+    };
+  };
+
+  //! Use a min heap to represent the list of candidate points.
+  //! We will use a vector and the std functions: push_heap() pop_heap().
+  //! We can not use a priority queue because we need to iterate over all the
+  //! candidates and std::priority_queue doesn't provide that interface.
+  typedef std::vector<Candidate> CandidateList;
```

Comment:
Do you think it would be better to use a `boost::priority_queue` here?  We use it in the `CosineTree` code for this exact reason.
http://www.boost.org/doc/libs/1_55_0/doc/html/heap.html